### PR TITLE
Add `boolean|int|float|decimal|string` union support for expr of string template

### LINKED ISSUE #29674: String template expression not Working for Union Types
**Description:**
Consider the following code:

```ballerina
public type Foo int|float|decimal|string|boolean;

public function main() {
    Foo foo = 4;
    string s = string`${foo}`;
}
```

This results in a compilation error: `incompatible types: expected '(int|float|decimal|string|boolean)', found '(int|float|decimal|string|boolean)'`

But using `${foo.toString()}` works.

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #29674 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/string/string-template-literal-negative.bal [hunk 1]
```diff
@@ -8,3 +8,15 @@ function stringTemplate2() returns (string) {
...
 }
+
+public type Foo int|float|decimal|string|boolean|();
+
+function tringTemplate3() returns string {
+    Foo foo = 4;
+    return string`${foo}`;
+}
+
+function tringTemplate4() returns string {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/string/string-template-literal-negative.bal:19]:
```suggestion
function stringTemplate4() returns string {
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/string/string-template-literal-negative.bal [hunk 2]
```diff
@@ -8,3 +8,15 @@ function stringTemplate2() returns (string) {
     string s = string `Hello ${name}`;
     return s;
 }
+
+public type Foo int|float|decimal|string|boolean|();
+
+function tringTemplate3() returns string {
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/string/string-template-literal-negative.bal:14]:
```suggestion
function stringTemplate3() returns string {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 3]
```diff
@@ -4622,6 +4622,19 @@ public boolean isUnionOfSimpleBasicTypes(BType type) {
         return isSimpleBasicType(type.tag);
     }
 
+    public boolean isUnionOfSimpleBasicTypesExceptNilType(BType type) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4625]: I think the accurate name now would be something like
```suggestion
    public boolean isNonNilSimpleBasicTypeOrString(BType type) {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 4]
```diff
@@ -6256,17 +6256,7 @@ private void checkStringTemplateExprs(List<? extends BLangExpression> exprs, boo
...
-                    if (type.tag != TypeTags.XML) {
-                        dlog.error(expr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES,
-                                BUnionType.create(null, symTable.intType, symTable.floatType,
-                                        symTable.decimalType, symTable.stringType,
-                                        symTable.booleanType, symTable.xmlType), type);
-                    }
-                    continue;
-                }
-
+            if (!types.isUnionOfSimpleBasicTypesExceptNilType(type)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6259]:
```suggestion
            if (!types.isNonNilSimpleBasicTypeOrString(type)) {
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 5]
```diff
@@ -4622,6 +4622,19 @@ public boolean isUnionOfSimpleBasicTypes(BType type) {
...
+        if (type.tag == TypeTags.UNION) {
+            Set<BType> memberTypes = ((BUnionType) type).getMemberTypes();
+            for (BType memType : memberTypes) {
+                if (memType.tag == TypeTags.NIL || !isSimpleBasicType(memType.tag)) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4629]: Shall we add a negative test where the RHS evaluates to true here?

e.g., `int[]|string[]`.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 4 above)
Comment (KRVPerera) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6259]: Cant, we use assignability here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 6]
```diff
@@ -6256,17 +6256,7 @@ private void checkStringTemplateExprs(List<? extends BLangExpression> exprs, boo
...
                 dlog.error(expr.pos, DiagnosticErrorCode.INCOMPATIBLE_TYPES,
```
Comment (KRVPerera) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6260]: Error message says `expected '(int|float|decimal|string|boolean)'` but actually we expect a subtype. Propose to have a new message.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 4 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6259]: Then we'll have to have a separate union type for this, right?

**Comment by MaryamZi** [comment]
Merging this to https://github.com/ballerina-platform/ballerina-lang/tree/lang-changes, will fix the suggestions in the PR to master.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 6 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:6260]: On second thoughts, this is like any other mismatched type error we have, right? WDYT?

```ballerina
function func(int|float|decimal|boolean|string x) {
}

public function main() {
    int[] x = [];
    func(x); // [ERROR] incompatible types: expected '(int|float|decimal|boolean|string)', found 'int[]'
}
```

**Pull request merged**